### ISSUE TITLE
Profiles as files vs profile as comment in thread

### ISSUE THREAD

**Issue by oscar-g** [open-source-society/computer-science#81]
May I ask:

What was the reasoning behind using a comment thread for student profiles vs using markdown files in the repository?

It seems like a file in a repository is more 'tangible'...because we don't actually issue any degrees of certificates, it seems like the 'academic program' would benefit from having students in their own separate files.  This way, we can link to a file or we can embed the file on a website; sure, the comments in an issue thread can be linked, but it's not as 'tangible'.

I realize that it's probably a lot of work to approve hundreds of pull requests...which is probably why we switched to comments.

That being said, I think managing your profile should be as easy as editing a comment.  Making a pull request is not very convenient for the student or the administrators.

I propose we develop a way to make a `{student-name}.md` file from each comment in the Student Profiles thread.  And, when someone updates their comment, their file is updated, accordingly.

What do you guys think?

**Comment by ericdouglas** [comment]
I think that we can use threads by now, and after we can create an application that lists all courses and allow students to create their profiles and track their progress. 

The student can also have a public profile and share it with others.

**Comment by oscar-g** [comment]
A web app for this also makes sense and is probably easier than using git hooks to make files and commits.

**Comment by oscar-g** [comment]
I think we could really benefit from an app that lets us manage our profiles.  A few features I think would be useful for the basic functionality:

* Login with github credentials
* Manage profile contact details
* Manage courses (maybe inprogress and completed)
* Ability to post (FB post, LinkedIn post, Twitter tweet) when you complete a course.

**Comment by ericdouglas** [comment]
@oscar-g really cool specs! We definitely should think in something like yours description :smile:

**Comment by dishark** [comment]
+1

@ericdouglas you can count on me if you wanna develop something like this.

**Comment by oscar-g** [comment]
@ericdouglas @dishark 

I think we should move with this as soon as we can. I am not sure if the project has other priorities, though.

Overall, the app looks to be really simple/straightforward.

One challenging aspect may be migrating the current profiles from the profile comment thread.

What do you think?

**Comment by odiefrom** [comment]
@oscar-g 

If you were to close the current application thread (thus freezing it), traversing it and scraping all the data into another format (database, objects, whatever) would be possible with the Github API:
https://developer.github.com/v3/issues/#get-a-single-issue

**Comment by oscar-g** [comment]
@odiefrom That sounds good...not too complicated, either.

We should be able to keep the thread open while developing the app. Then, the migration would take place when we're ready to launch the site.

**Comment by odiefrom** [comment]
@oscar-g Right...only fire off the migration (and thus, closing the thread) once the app is launched (unless you care about registration order / internal numbering of students, in which case you have to tweak that timing slightly).

**Comment by dishark** [comment]
Cool @oscar-g, 

waiting @ericdouglas's feedback

**Comment by ericdouglas** [comment]
Hey, guys!

Such app will be really helpful, and indeed we have so many cool things that we can add to it.

We need to keep a very lean mind here, and focus on our MVP.

Our MVP now consists in finish our curriculum with possible missing topics/courses.

After we have a "frozen version" of our curriculum, we should start to think in that application.

It makes sense for you?

**Comment by ericdouglas** [comment]
There are no problem with students enrolling in the issue `#31`. We can easily ask for them to register in our app, if it is the case.

**Comment by dishark** [comment]
Yeah importing data isn't big deal.

What about an app you register with github. Point where you are from on google's map. And check courses out as you go.

__Note:__ We need to take care, 'cause we might end up losing the easy changing feature that github gives to us.

**Comment by ericdouglas** [comment]
We need to be very careful here and not disturb the student experience with some moving parts.

Let's first understand our business and see what will be our necessities.

The main point here is to enjoy the path, and learn a lot =)

**Comment by oscar-g** [comment]
Good points, @ericdouglas.  We should take care not to release something that hasn't been thoroughly planned and tested.

Right now, I think the google group is a decent place for members to discuss...and github works well for some things.  With time, we can work to add more sophisticated conversation options, like a topic=-based discussion forum.

Also, I think members could benefit from being able to display their profile in a single page, where the focus of that page is one member's contact information and achievements.  As I've said before, I think having a single member profile spaces could bolster the authenticity and (dare I say) authority of the program.  In a way, the design and marketing of the effort will create a sense of higher organization that will look great when a member shows off their work.  

Maybe we're not trying to come off as a degree from an established university (at least that's not the way I think of this, but I am open to hearing your thoughts on this and learning more about the philosophy of OSS)...the fact is that that members learn skills they can use in the real world and complete a self-guided curriculum, something that not everyone can do because it requires commitment and patience (in some cases, even more discipline than a university degree because of jobs, kids, and living situations).  I think this is a great quality of folks going through self-guided educational programs.  For this reason, I think it will be important to give members a space to show their achievements in an 'official' manner.

Nevertheless, let's keep planning this!

**Comment by ericdouglas** [comment]
@oscar-g I agree with all of your arguments. Let's plan this and deliver an awesome app that attends all of our necessities.

It'll definitely be a big plus for all OSS students! :smile:

**Comment by maicoqb** [comment]
The profile as comments approach is more simple to deal now.
But the app thing is a great idea.

**Comment by thebergamo** [comment]
Hi guys!

I really love this project is for me the best project in github for learn! Congrats @ericdouglas!

Well, just put here my ten cents.

When I need update my course profiles in comments issue are really bad and untrackable for perform new users or any way because when we have a lot of students imagine you update your comment on the 8k position? 

Maybe for this case, two approaches can be more useful, the forks maybe the user can use the "Students Register" an repo when the user put yours study objects, and if I need se how many students the project have we can simple see the forks and see yours repos.

Or using a simple webapp to support this. 

Thanks and grow up!

**Comment by ericdouglas** [comment]
@thebergamo many thanks for your suggestions and your feedback!

Any solution involving github seems not scalable. I think that a web/mobile app with [those](https://github.com/open-source-society/computer-science/issues/81#issuecomment-145606663) specifications by @oscar-g will be a really good solution.

Students will have to use the search from browsers to find their profile, because we already have a **lot** of students :smile:

**Comment by oscar-g** [comment]
If I may suggest WordPress as a platform.
We could extend the built-in User object and add the data fields we need; we can also build a user search function.  If we want to get fancy, we could add a "Course" object that will let us manage the available courses like posts...then users can choose from this list of courses.

Any thoughts on using WordPress for this?

**Comment by ericdouglas** [comment]
I can't help with WordPress... But let's see what other guys have to say about the stack.

**Comment by SoullessWaffle** [comment]
I'm no expert, but I've heard some very good things about [Meteor](https://www.meteor.com/).

More info: http://wiki.dandascalescu.com/essays/why_meteor

**Comment by oscar-g** [comment]
One of the reasons I suggested WP is that the content management code is already written for us, which will things easier to implement. Another cms-type platform we could use could be http://keystonejs.com/

@SoullessWaffle I've heard good things about Meteor, too.  I looked at a tutorial about a year ago...I'm sure a lot has changed since then.  I do remember that Meteor excels at real-time data and syncing changes in the data to all connected clients.  I wonder if that is something that will be useful for the type of application we're designing?  Maybe not in the 'member profile' aspect, but if we implement discussions, real-time data updates would be very nice.

**Comment by Heasummn** [comment]
I think Django might be nice, the github API wrapper is also really easy to use. Rails and Meteor would be nice if we want to make a nice API for mobile apps, but that can be done in Django as well. Spring and Boot are too big for something like this, but Flask could also be viable be as a backend.

**Comment by waterlooSunset** [comment]
I think that each of the stacks suggested so far have their strengths and weaknesses. It seems from this issue thread that we are all tending toward favoring an app in the near future. Perhaps we should SWOT each of the currently and future proposed stacks to help decide what direction we go in. I'm mostly a javascript type, but I'd love to contribute to the build out in any way I can regardless of the stack ultimately chosen!

**Comment by thebergamo** [comment]
Wow, well the web/mobile app is the best fit really. 
[VersoApp](http://versoapp.com/) is an platform for learning built in Meteor. 
But if we will develop the platform, it is a huge work, but we can do that as quickly we want. In my particular case opt to develop the entire app using just Node.js, and not a complete framework. Because Node.js is simple and any people who knows JS can help in the project.

**Comment by ericdouglas** [comment]
I'm with @thebergamo . I would suggest to use Node.js (Express), MongoDB and maybe React, as we'll have a lot of cards to display on the screen.

**Comment by oscar-g** [comment]
@ericdouglas, that might be the best choice. Like @thebergamo said, most people will be able to contribute to a javascript stack.

I've been messing around with ReactJS and there is a steep learning curve, but it is rewarding.

If we do start something from scratch, let's try not to reinvent the wheel.  One of the strengths of using a framework with a cms is that an an administrator could easily manage profile content, class offerings, and pages...so this app could double as the homepage for the curriculum. Also, imporant things like login security will come standard with a cms.

**Comment by SoullessWaffle** [comment]
@ericdouglas Meteor runs on Node.js (meaning you can `require` node modules), it comes with [full MongoDB integration by default including latency compensation](https://www.meteor.com/full-stack-db-drivers) and also [lets you use react](https://www.meteor.com/tutorials/react/creating-an-app).
A mobile app would be great as well, and [Meteor uses Cordova for this](http://www.sitepoint.com/beginners-guide-mobile-development-meteor/).

I agree with @waterlooSunset in that an evaluation of each of the stacks' strengths and weaknesses is a good idea.

To this end, I've created [a spreadsheet](https://docs.google.com/spreadsheets/d/1bi-wtzCaWU9jXx3y6TTJhxwgcM-aoRnsUa0zF3IGTQc/edit?usp=sharing) that anyone can edit.
I've already started populating it, but I don't know enough about the other stacks to complete the whole sheet. Feel free to add more stacks and features to it.

**Comment by ericdouglas** [comment]
@SoullessWaffle thanks for the intiative! It'll be very helpful.

**Comment by oscar-g** [comment]
This is great @SoullessWaffle, we've got a wide range of options.

**Comment by maicoqb** [comment]
And after options? We will vote?

**Comment by maicoqb** [comment]
Hey guys the Profile as Comment is going bad.

#31 
>	@ericdouglas opened this issue 3 days ago · 683 comments

And growing...

**Comment by ericdouglas** [comment]
@maicoqb it's true... maybe open another issue just to give some time for us. What do you think?

Students Profile 2

**Comment by oscar-g** [comment]
Let's discuss the options for a day or two...then vote somehow?

**Comment by ericdouglas** [comment]
@maicoqb @oscar-g yes, I think that we can choose tomorrow.

**Comment by oscar-g** [comment]
Almost 700 in that members thread...that's awesome! and a dangerous way to organize heh

**Comment by ericdouglas** [comment]
Do you know if a lock the issue people will sitll be able to edit their comments?

**Comment by ericdouglas** [comment]
@oscar-g it's indeed an amazing number! In just 3 days :scream:

**Comment by oscar-g** [comment]
@ericdouglas, try locking this thread to test...you can open it back up, I think

**Comment by ericdouglas** [comment]
@oscar-g I made this test... When a issue is locked, nobody can edit a comment.

**Comment by maicoqb** [comment]
Ouch, then leave open the Students Profile 2 and close the issue #31.
I know that closed issues can be edited.

**Comment by maicoqb** [comment]
But closed issues can also be commented. :confused:

**Comment by SoullessWaffle** [comment]
Yeah, closing the issue wouldn't do much except confuse would-be students.

**Comment by vongst** [comment]
Hacked together something simple here as proof of concept:

http://104.236.1.56/public/#/courses/
http://104.236.1.56/public/#/student/12
http://104.236.1.56/public/#/students/

Any AngularJS/PHP devs wanna take this up?

**Comment by thebergamo** [comment]
@vongst thats very pretty!

The source has in github?

**Comment by oscar-g** [comment]
@vongst, Great start! Maybe add it the 'dev' branch on your fork of this
repo.  Then we can make pull requests against your branch.

**Comment by vongst** [comment]
@thebergamo 
@oscar-g 
@ericdouglas 

Thanks guys, here's the repo: https://github.com/vongst/open-source-society

Take it away! It's still very rough around the edges, if there's official interest/support I could try spending more time on it

**Comment by mseyne** [comment]
That a nice job ! A web site for opensourcesociety would help many people (to discover) to join as a first step before to jump in github. I don't know if I can help for content (don't know php) but I will, thanks for that.

**Comment by SoullessWaffle** [comment]
@vongst that does look great! And nice job with the data scraping, looks like it was less of an issue than I initially thought.
I'm in the same boat as @mseyne in that I don't know PHP, so I doubt I'll be able to help with the code.

**Comment by ericdouglas** [comment]
I don't know PHP too...

We should decide which stack we'll use.

@oscar-g @thebergamo @dishark @odiefrom @maicoqb @Heasummn @waterlooSunset @SoullessWaffle @vongst 

Which technologies are you able to help?

I can help with Node.js, MongoDB and Angular.js.

**Comment by SoullessWaffle** [comment]
I can help with Node.js, MongoDB, CSS & HTML (also Stylus & Jade), as well as Meteor.

I think by now it's pretty clear I'm enthusiastic about Meteor.
One of the nice things about Meteor is that if you know Node.js, or even just JavaScript, it should not be difficult to learn since it runs on those technologies.
There's a talk I found which serves as a great introduction to Meteor and what it can do: https://youtu.be/SYqyWff6iMQ?t=217
If you're busy, here's (an outdated) 10 minute video: https://www.youtube.com/watch?v=fsi0aJ9yr2o

**Comment by ericdouglas** [comment]
@SoullessWaffle 

I'm a bit against monolithic solutions, but if there are more people that know Meteor, and this number is greater than the people that know another stack, we can go with Meteor for sure.

**Comment by oscar-g** [comment]
I can help with any of JS, PHP, MySQL, Mongo, and HTML/CSS.  I know Angular
well and some ReactJS.

It looks like most of us know js.

@vongst started on the right track. Maybe we can split into frontend (JS)
and backend (PHP) teams.

Let's also try to get the requirements ready, before starting to code too
much.

**Comment by ericdouglas** [comment]
@oscar-g specifications are a **must have**! :smile:

Maybe we can try to document it using [app-specs](https://github.com/ericdouglas/app-specs), so we'll have two more projects.

I also think that we need to have two team, and communicate those apps via a REST API.

**Comment by mseyne** [comment]
I agree with @oscar-g we should split between frontend (JS) and backend (PHP) teams. Though, for backend I would vote for python/django because it is what I am learning (and I love it) but I am not a veteran enough to push for this choice. Hope python coders will come here :)

**Comment by SoullessWaffle** [comment]
@ericdouglas fair enough, all I'm saying is it might be the right kind of monolith.
If we're not using Meteor, then REST sounds good. There's also websockets if we want realtime, perhaps through [Primus](https://github.com/primus/primus).

**Comment by thebergamo** [comment]
You talk a lot when I'm not here :smile: 

@vongst I can help with PHP, but now isn't my expertise. If we can work with JS(Node.js) to creating a REST API is an profit solution because more people can participate, JS is a common language.
PHP is so common in Web too  :p
By the way, you has do a great job! Congratz!!

@oscar-g  the ideia to split the groups in Backend and Frontend is the best approach I think. 
@ericdouglas You can create 2 repos for the App? A repo for frontend and other to backend and centralising the discussion?

**Comment by odiefrom** [comment]
@ericdouglas  Sorry to be late to the party.  I can help with PHP and SQL (more geared toward MySQL, but I don't see the differences coming up in a project of this scope). :)

**Comment by ericdouglas** [comment]
@SoullessWaffle yeah! Meteor here would be a good fit!

@thebergamo for sure!

For those that have interest to become a member of the OSS organization, just let me know! :smile: 

I would love to have more contributors here!

**Comment by oscar-g** [comment]
Maybe we should establish requirements before we choose a stack?

**Comment by ericdouglas** [comment]
# Backend

:octocat: |Node.js | PHP | Python | MongoDB | MySQL
:--|:--:|:--:|:--:|:--:|:--:
@oscar-g | |✔| |✔|✔
@thebergamo | ✔|✔ | |✔|✔
@dishark | | | ||
@odiefrom | |✔ | ||✔
@maicoqb | | | ||
@Heasummn | | | ||
@waterlooSunset | | | ||
@SoullessWaffle | ✔| | |✔|
@vongst | |✔ | ||
@eric | ✔| | |✔|

# Frontend

:octocat:|Angular.js | React | HTML | CSS | Jade | Stylus | Sass
:--|:--|:--:|:--:|:--:|:--:|:--:|:--:
@oscar-g | ✔| ✔| | | | |
@thebergamo | | | | | ||
@dishark | | | | | ||
@odiefrom | | | | | ||
@maicoqb | | | | | ||
@Heasummn | | | | | ||
@waterlooSunset | | | | | ||
@SoullessWaffle | | |✔ |✔ |✔ |✔|
@vongst |✔ | |✔ | ✔| ||
@eric | ✔| | | | ||

# UX/UI

??? :sob:

**Comment by oscar-g** [comment]
@ericdouglas I'm able to help out with any of these.

As far as features go:
* Login with github credentials
* Manage profile contact details (including map position)
* Manage courses (maybe inprogress and completed)
* Ability to post (FB post, LinkedIn post, Twitter tweet) when you complete a course.
* admins can manage (CRUD) courses and profiles
* list all profiles, search profiles

On the front page, maybe we can show a map with the user map positions and when you click the marker it takes you to the profile page.

What else should we include?

**Comment by m-schrepel** [comment]
I am also a big fan of meteor, but I can help out with Node, HTML, CSS, Jade, React, and MongoDB

Of course, I'd enjoy using React with Meteor just as well.

**Comment by SoullessWaffle** [comment]
@ericdouglas Nice touch with the octocat :+1: 
Since Node.js and Meteor kinda blur the line between Frontend and Backend, here are some tables sorted by Languages and Frameworks:

# Languages
:octocat: | HTML | CSS | JS | PHP | SQL
---------- | -------- | ------ |---- | ----- | ----
@oscar-g |  |  | ✔ | ✔ | ✔ |  
@thebergamo |  |  | ✔ | ✔ | ✔ |  
@dishark |  |  |  |  |  |  
@odiefrom |  |  |  | ✔ | ✔ |  
@maicoqb |  |  |  |  |  |  
@Heasummn |  |  |  |  |  |  
@waterlooSunset |  |  |  |  |  |  
@SoullessWaffle | ✔ | ✔ | ✔ |  |  |  
@vongst | ✔ | ✔ | ✔ | ✔ |  |  
@ericdouglas |  |  | ✔ |  |  |  
@m-schrepel | ✔ | ✔ | ✔ |  |  |  

# Frameworks
:octocat: | Angular.js | Meteor | Node.js | React
---------- | -------- | ------ |---- | -----
@oscar-g | ✔ |  |  | ✔ |  
@thebergamo |  |  | ✔ |  |  
@dishark |  |  |  |  |  
@odiefrom |  |  |  |  |  
@maicoqb |  |  |  |  |  
@Heasummn |  |  |  |  |  
@waterlooSunset |  |  |  |  |  
@SoullessWaffle |  | ✔ | ✔ |  |  
@vongst | ✔ |  |  |  |  
@ericdouglas | ✔ |  | ✔ |  |  
@m-schrepel |  | ✔ | ✔ | ✔ |  

And also a table for CSS preprocessors, as we will probably want to pick one (for code reusability):

# CSS and its Preprocessors
:octocat: | CSS | Less | Sass | Stylus | Swith | Turbine
---------- | ------ | ------ | ------ | -------- | ------- | ----------
@oscar-g |  |  |  |  |  |  |  
@thebergamo |  |  |  |  |  |  |  
@dishark |  |  |  |  |  |  |  
@odiefrom |  |  |  |  |  |  |  
@maicoqb |  |  |  |  |  |  |  
@Heasummn |  |  |  |  |  |  |  
@waterlooSunset |  |  |  |  |  |  |  
@SoullessWaffle | ✔ |  |  | ✔ |  |  |  
@vongst | ✔ |  |  |  |  |  |  
@ericdouglas |  |  |  |  |  |  |  
@m-schrepel | ✔ |  |  |  |  |  |  

*Please let me know if I missed anything*

**Comment by ericdouglas** [comment]
@SoullessWaffle great! Thanks for that!

Just MongoDB is not listed.

**Comment by SoullessWaffle** [comment]
@ericdouglas good catch! I'll add it under frameworks.

**Comment by ericdouglas** [comment]
@SoullessWaffle :+1:

**Comment by SoullessWaffle** [comment]
@dishark @odiefrom @maicoqb @Heasummn @waterlooSunset are you guys familiar with JS? If so I'll add that to the tables.

**Comment by odiefrom** [comment]
@SoullessWaffle  I have basic JS practice, nothing with frameworks or anything.  I've also been working with HTML5 features lately, and have been using HTML in general for about a decade, so throw me down for that if I can help :)

**Comment by jdinartejesus** [comment]
@SoullessWaffle maybe also could help you guys, I know some React, Stylus or Sass, Flux and Reflux. I liked the project, I would love to contribute.

**Comment by SoullessWaffle** [comment]
@odiefrom @jdinartejesus I've added your skills to the tables :)
I guess it's up to @ericdouglas to handle the invites to the OSS org.

**Comment by waterlooSunset** [comment]
@soulesswaffle I've got some Javascript experience and some html5, css3 and jade experience. A liitle bit of exposure to passport, express, mongoose library/frameworks. Oh and little jQuery and bootstrap, grunt and mocha. I'd love to help anyway i can even its just writing unit tests

**Comment by thebergamo** [comment]
@SoullessWaffle can add to me HTML and CSS I can help in frontend too =D In frameworks I can all at all too!

**Comment by SoullessWaffle** [comment]
@waterlooSunset @thebergamo I've added you both :)

@ericdouglas for reference, here's a realtime production-stage collaboration app built with Meteor: https://hansoftx.com
They also wrapped the app in Cordova so it's available on Android/iOS.

**Comment by jdinartejesus** [comment]
@SoullessWaffle  don't forget also to select HTML for my side :+1:

**Comment by ericdouglas** [comment]
You're **amazing**, guys! Thank you so much for your help!

I sent an invite for all of you :wink: 

I created two repositories, [ossu-api](https://github.com/open-source-society/ossu-api) (that will contain our backend code) and [ossu-ui](https://github.com/open-source-society/ossu-ui) (that will contain our frontend code).

As I can see now, our stack will be:

- Node.js
- MongoDB
- Angular or React (the frontend team should decide)
- Stylus
- HTML

Now we need to split the teams!

Let's make history here! :rocket:

**Comment by jdinartejesus** [comment]
Great Job guys!

Where should we discuss about frontend tools?
I don't understand that much about backend, so I guess already know my team! 

Really excited about the project and learn CS at the same time! :+1:

**Comment by ericdouglas** [comment]
@jdinartejesus I think that you can open an issue in the ossu-ui repository and start to discuss about it :smile:

**Comment by jdinartejesus** [comment]
Already done @ericdouglas.

How want to join the front-end, please give your feedback and help us choose the best approach to everyone. 

Here is the issue, feel free to give your opinion and recommendation.
https://github.com/open-source-society/ossu-ui/issues

Let's change the education! :book:

**Comment by SoullessWaffle** [comment]
Looks like it's time for more tables! :smile: 

# Teams
:octocat: | Front-end | Back-end
---------- | ------------- | ------------
@oscar-g |  |  | 
@thebergamo |  |  | 
@dishark |  |  | 
@odiefrom |  |  | 
@maicoqb |  |  | 
@Heasummn |  |  | 
@waterlooSunset |  |  | 
@SoullessWaffle | ✔ |  | 
@vongst |  |  | 
@ericdouglas |  |  | 
@m-schrepel |  |  | 
@jdinartejesus  | ✔ |  | 

_If you would like to collaborate in this project, please let us know whether you want to be part of the Front-end or Back-end team_

**Comment by evanplaice** [comment]
@ericdouglas Why 2 teams? This isn't exactly a large undertaking, unless the data api is meant to be used for both web and mobile apps in the future.

The important questions **should be** what is the scope?
- A searchable list of users
- A list of courses with attendees (in progress?/complete?)

Is that it?
If so, it might be a good idea just to to use the layout provided by @vongst and scrape user listings daily to update the static version. If you use Angular, you can dump the updated user listing as a json file and read it in as a data variable. That'll make it easy to implement search.

If the desire is to grow into a bigger app/platform, where are the webserver and database going to be located. Will development servers be made available for testing? Who will be responsible for the setup, payment, maintenance?

Will this be used to host relevant documentation in the future? It wouldn't be too difficult to add Markdown directive support in angular. That way, documentation can be written and managed directly in source control without requiring a back-end or ACL to edit.

BTW, you can put me down for back-end dev.

**Comment by SoullessWaffle** [comment]
Two questions:
- Do we want to use a REST api or websockets for client-server communication (or a mix of both)?
 - The devs in this project may have more experience using REST than websockets
 - Websockets are better suited for realtime apps because there is less overhead since the socket doesn't close after every message
- For the front-end, do we want to use [Jade](http://jade-lang.com/) or just HTML?
 - Jade will be quite a bit faster to develop in for anyone who has ever used it
 - Jade's syntax is easy to learn (HTML without angle brackets and with some javascript-like syntax)
 - Everyone knows HTML
 - Some (understandably) can't be bothered to learn a new syntax if they can instead use what is familiar (HTML)

**Comment by evanplaice** [comment]
@SoullessWaffle 

**Back-End**
Definitely REST. Websockets would only be necessary if there was some form of real-time communication such as chat. Either way, REST would be used for the content portion.

**Front-End**
If Angular is the chosen framework, then there's no need for Jade.

I may be the outlier here but I prefer not to include additional framework layers until I can justify the added complexity.

@ericdouglas How will commits be handled. Will direct push access be granted or is it back to the ol' fork, branch, submit, merge churn?

**Comment by SoullessWaffle** [comment]
@evanplaice it seems that @ericdouglas has already given me push access to both repos, so I guess that's what we're doing.

With Angular2 just around the corner, I doubt many would find learning Angular1 very enticing.
Having used neither, if we have to choose between one or the other I would say React sounds like a better deal.

**Comment by jdinartejesus** [comment]
@evanplaice I also would recommend Reactjs, also see this project as a opportunity to learn. 

So if we had thins like Jade or other simple tools easy to learn will be good for people start implementing them and learn in the end.

**Comment by ericdouglas** [comment]
@evanplaice 

> "Why 2 teams? This isn't exactly a large undertaking, unless the data api is meant to be used for both web and mobile apps in the future."

We'll create a web and mobile app :wink: 

> "The important questions should be what is the scope?"

Basically, [these](https://github.com/open-source-society/computer-science/issues/81#issuecomment-146962094) are the specs.

Students need to register again, but it'll be for a good cause. I think that no one will care about it.

> "Will development servers be made available for testing? Who will be responsible for the setup, payment, maintenance?"

We need to discuss about it.

> "Will this be used to host relevant documentation in the future?"

To document what kind of things?

> "How will commits be handled. Will direct push access be granted or is it back to the ol' fork, branch, submit, merge churn?"

I'd highly recommend use [this](http://nvie.com/posts/a-successful-git-branching-model/) [git workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow/).

@SoullessWaffle 

> "Do we want to use a REST api or websockets for client-server communication (or a mix of both)?"

REST. We need basically a CRUD system for now. :smile:

**Comment by jdinartejesus** [comment]
I agree with @ericdouglas, also I recommend use git workflow. Good way to keep organize.

**Comment by evanplaice** [comment]
@ericdouglas OK, I switched to committing to the 'develop' branch. Also pushed another pull request. There's a minor conflict in README.md.

Getting a basic database up with CRUD access should be straightforward enough. MongoDB isn't a relational database so faking relationships between collections is pointless. I'll see if I can come up with a basic data model to share.

GitHub login is going to be a huge pain and is beyond my ability. oAuth is not the easiest authentication method to implement. Hopefully, somebody with prior experience with it will come along.

If the site is handling logins, HTTPS support would be ideal for security purposes.

Posting to social media when you complete a course is strictly a front-end thing. Once a user is viewing his/her profile, all the social links should be there. Authentication to post should come up when the social site's API is called.

User map positions will only be possible if location data is made available. Unless the users type a location into their profile, then a dataset that maps city names to coordinates is needed.

**Comment by SoullessWaffle** [comment]
> oAuth is not the easiest authentication method to implement.

\*cough\* [Meteor](https://youtu.be/SYqyWff6iMQ?t=1592) \*cough\*
Alright, I'll shut up now ( ͡~ ͜ʖ ͡°)

**Comment by SoullessWaffle** [comment]
So, to recap:

# Project Tools

Function | Tool
----------- | ----------
Team Communication | [Slack][slack]
Development Planning | [Hansoft X][hansoftx]
Git Workflow | [Git Flow][gitflow]
Server | [Node.js][nodejs]
Database | [MongoDB][mongodb]
Asset Bundler | [Webpack][webpack]
App Framework | [ReactJS][reactjs]
View Language | HTML
CSS Preprocessor | [Stylus][stylus]
CSS Framework | [Semantic UI][semanticui]
Unit Testing | [Jest][jest]?
Code Coverage | [Coveralls][coveralls]?
CI Service | [Travis CI][travisci]?
Hosting | [DigitalOcean][digitalocean]*

\* I have some credits on DigitalOcean, and depending on how long it takes it should be enough to cover development.

[slack]: https://slack.com
[hansoftx]: https://hansoftx.com
[gitflow]: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow
[nodejs]: https://nodejs.org
[mongodb]: https://docs.mongodb.org/manual/
[webpack]: https://webpack.github.io/
[reactjs]: https://facebook.github.io/react/
[stylus]: https://learnboost.github.io/stylus/
[semanticui]: http://semantic-ui.com/
[jest]: https://facebook.github.io/jest/
[coveralls]: https://coveralls.io/
[travisci]: https://travis-ci.org/
[digitalocean]: https://www.digitalocean.com/

**Comment by ericdouglas** [comment]
@SoullessWaffle I'd suggest

- Transpiler: Babel (backend & frontend)
- Test (backend): Mocha, Chai, Sinon.js, Supertest
- Code Coverage: [codecov.io](https://codecov.io)
- CI Service: Travis :+1: 
- Hosting: DO* :+1:

\* I've some credits, too.

**Comment by ericdouglas** [comment]
[Form](https://ericdouglas.typeform.com/to/ZmyDRd) to catch your e-mail and send an invitation to our Slack Channel for the dev team.

**Comment by jdinartejesus** [comment]
@ShirazSuleman Looks good for me. :+1: 

For testing maybe using Jest for React it's better, if i'm not wrong. @ericdouglas

**Comment by ericdouglas** [comment]
@jdinartejesus those tools are for the backend :smile:

**Comment by SoullessWaffle** [comment]
@oscar-g	@thebergamo	@dishark	@odiefrom @maicoqb @Heasummn @waterlooSunset @vongst @m-schrepel @jdinartejesus @evanplaice 

If you want to contribute, [please fill out this form][form] so we can send you invites to the [Slack][slack] chat and [Hansoft X][hsx] development board.

[form]: https://soullesswaffle.typeform.com/to/xuTU4O
[slack]: https://slack.com
[hsx]: https://hansoftx.com

**Comment by SoullessWaffle** [comment]
@ericdouglas that table I made lists all the tools we're using for the project, including back-end and front-end.

**Comment by oscar-g** [comment]
Looks great, everyone! Let's keep planning in the development board.

**Comment by evanplaice** [comment]
@oscar-g I filled out the form... Still awaiting approval.

**Comment by jdinartejesus** [comment]
I didn't receive anything yet about Hansoft x or Slack. I'm the only one? @ericdouglas

**Comment by SoullessWaffle** [comment]
@mseyne @waterlooSunset @oscar-g @jdinartejesus I invited you guys to the dev board. You should receive an email any minute.

@evanplaice we didn't receive your form submission, could you please [try again][form]?

[form]: https://soullesswaffle.typeform.com/to/xuTU4O

**Comment by ericdouglas** [comment]
@jdinartejesus I sent the slack's invitation for you. 

@evanplaice we didn't receive your info... Can you fill this out again?

**Comment by jbhum** [comment]
I am super excited about this! I just submitted the form and I am excited to participate and help wherever I can.

**Comment by ericdouglas** [comment]
@jbhum great! Thank your for your interest! :smile: 

Closing this issue because we already started to develop the app.

**Comment by SoullessWaffle** [comment]
If anyone else wants to help develop the app, please see [these instructions](https://github.com/open-source-society/ossu-ui/blob/develop/CONTRIBUTING.md).

**Comment by evanplaice** [comment]
@SoullessWaffle Done...

**Comment by vongst** [comment]
@SoullessWaffle sorry I just saw this and the form's closed. Can I get an invite to slack?

**Comment by evanplaice** [comment]
@vongst You should have an invite from @SoullessWaffle.

**Comment by ShirazSuleman** [comment]
I filled in the form but didn't receive a slack invite, would it be possible for me to get one?

**Comment by jbhum** [comment]
@ShirazSuleman You will probably get one some time tomorrow. Most of the people that can send one have not been online for most of the day. Post on here again tomorrow evening if you haven't received one yet and I will make sure someone sends one your way.

**Comment by firstdoit** [comment]
Hi! Would love to help on all things Node and React!

**Comment by jbhum** [comment]
@firstdoit Feel free to look at [this](https://github.com/open-source-society/ossu-ui/blob/develop/CONTRIBUTING.md) to get started contributing to this project!

**Comment by firstdoit** [comment]
Great! Can someone invite me to slack at gadr90@gmail.com ?

**Comment by maicoqb** [comment]
Invite me to slack too, maicoqb@gmail.com

**Comment by ericdouglas** [comment]
@maicoqb see [this](https://github.com/open-source-society/ossu-ui/blob/develop/CONTRIBUTING.md#membership) section. :wink:

**Comment by maicoqb** [comment]
@ericdouglas Done. ;)

**Comment by ericdouglas** [comment]
@maicoqb :+1:

**Comment by ShirazSuleman** [comment]
@jbhum I've not received any slack invite yet.

**Comment by jbhum** [comment]
@ShirazSuleman did you fill out the form on the page I linked?

**Comment by ShirazSuleman** [comment]
Yep, I can do it again if needed?
